### PR TITLE
nonMasqueradeCIDRs configuration added to the helm chart

### PR DESCRIPTION
**Pull request by cyclinder**
Signed-off-by: cyclinder qifeng.guo@daocloud.io

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

Fixes: #[20051](https://github.com/cilium/cilium/issues/20051)

```release-note
add  `nonMasqueradeCIDRs` configuration to the ipMasqAgent section in Helm Chart values.
```

### CODE REVIEW

File: install/kubernetes/cilium/templates/cilium-configmap.yaml [hunk 1]
```diff
@@ -454,6 +454,9 @@ data:
 {{- end }}
 {{- if and .Values.ipMasqAgent .Values.ipMasqAgent.enabled }}
   enable-ip-masq-agent: "true"
+  config:
```
Comment (brb) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml]: Have you tested your change? I don't see how this `config` can be propagated as the ip-masq-agent setting to cilium-agent.

Comment (cyclinder) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml]: uh i miss something..
```
- name: ip-masq-agent
        configMap:
          name: cilium-config  # change it to cilium-config
          optional: true
          items:
          - key: config
            path: ip-masq-agent
```
I made some test for helm install, but the agent pod failed to init, No useful information could be found from either `kubelet` or `kubectl describe`

```
[root@master cilium]# helm install cilium . --set image.tag=v1.12.0-rc2  --set operator.image.tag=v1.12.0-rc2
NAME: cilium
LAST DEPLOYED: Fri Jun 10 20:01:55 2022
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
You have successfully installed Cilium with Hubble.

Your release version is 1.11.90.

For any further help, visit https://docs.cilium.io/en/v1.11/gettinghelp
[root@master cilium]# kubectl get po
NAME                               READY   STATUS       RESTARTS     AGE
cilium-cfvsk                       0/1     Init:Error   0            4s
cilium-gt2q9                       0/1     Init:Error   1 (3s ago)   4s
cilium-operator-84cbdd8b5c-dbr87   1/1     Running      0            4s
cilium-operator-84cbdd8b5c-jq7pf   1/1     Running      0            4s
```
```
Events:
  Type     Reason     Age                From               Message
  ----     ------     ----               ----               -------
  Normal   Scheduled  33s                default-scheduler  Successfully assigned default/cilium-cfvsk to master
  Normal   Pulled     32s                kubelet            Successfully pulled image "quay.io/cilium/cilium:v1.12.0-rc2" in 987.115567ms
  Normal   Pulled     30s                kubelet            Successfully pulled image "quay.io/cilium/cilium:v1.12.0-rc2" in 1.007922246s
  Normal   Pulling    13s (x3 over 33s)  kubelet            Pulling image "quay.io/cilium/cilium:v1.12.0-rc2"
  Normal   Created    12s (x3 over 32s)  kubelet            Created container mount-cgroup
  Normal   Started    12s (x3 over 32s)  kubelet            Started container mount-cgroup
  Warning  BackOff    12s (x3 over 29s)  kubelet            Back-off restarting failed container
  Normal   Pulled     12s                kubelet            Successfully pulled image "quay.io/cilium/cilium:v1.12.0-rc2" in 1.064636306s
```
This seems to no matter with `ipMasqAgent`, enable it  or disable  has the same result..

File: install/kubernetes/cilium/templates/cilium-configmap.yaml [hunk 2]
```diff
@@ -454,6 +454,8 @@ data:
 {{- end }}
 {{- if and .Values.ipMasqAgent .Values.ipMasqAgent.enabled }}
   enable-ip-masq-agent: "true"
+  config: |-
```
Comment (joestringer) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml]: I think this config should be somehow specific to IP Masq agent, naming it just `config` will be confusing .

File: install/kubernetes/cilium/values.yaml [hunk 3]
```diff
@@ -1076,6 +1076,12 @@ ipam:
 # -- Configure the eBPF-based ip-masq-agent
 ipMasqAgent:
   enabled: false
+  config:
```
Comment (joestringer) [inline install/kubernetes/cilium/values.yaml]: After #20066 was merged, now these changes need to be made to `install/kubernetes/cilium/values.yaml.tmpl` instead, then run `make -C install/kubernetes` to update this file.

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml [hunk 4]
```diff
@@ -699,7 +699,7 @@ spec:
       {{- if and .Values.ipMasqAgent .Values.ipMasqAgent.enabled }}
       - name: ip-masq-agent
         configMap:
-          name: ip-masq-agent
+          name: cilium-config
```
Comment (joestringer) [inline install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml]: I was under the impression that the way this works today is that Cilium can take whatever existing ip-masq-agent config you have (used even with non-Cilium ipmasq-agent), and read, understand and implement it. Is that not the case today?

**Review by joestringer — CHANGES_REQUESTED** [review]

File: install/kubernetes/cilium/templates/cilium-configmap.yaml (hunk 2 above)
Comment (cyclinder) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml]: Thanks! @joestringer 

Maybe change `config` to `ip-masq-agnet-config` something like?

File: install/kubernetes/cilium/values.yaml (hunk 3 above)
Comment (cyclinder) [reply install/kubernetes/cilium/values.yaml]: Okay

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml (hunk 4 above)
Comment (cyclinder) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml]: Thanks! 

Do you mean we need to create a `configmap` with the name "ip-masq-agent" here, and then we refer to it here?

Comment (joestringer) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml]: Yes, that's my understanding. This line should not need to change at all. As far as I know, you just need to create the ip-masq-agent configuration configmap in the cluster.

File: install/kubernetes/cilium/values.yaml [hunk 5]
```diff
@@ -1076,6 +1076,12 @@ ipam:
...
+    nonMasqueradeCIDRs:
+    - 10.0.0.0/8
+    - 172.16.0.0/12
+    - 192.168.0.0/16
```
Comment (joestringer) [inline install/kubernetes/cilium/values.yaml]: I don't know if we can ship these by default. The network may or may not guarantee that bidirectional connectivity between pod IPs and these IP ranges will work. These need to be configured by the user.

File: install/kubernetes/cilium/values.yaml (hunk 3 above)
Comment (cyclinder) [reply install/kubernetes/cilium/values.yaml]: Thanks, Done

File: install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml (hunk 4 above)
Comment (cyclinder) [reply install/kubernetes/cilium/templates/cilium-agent/daemonset.yaml]: Agree! Thanks.

File: install/kubernetes/cilium/values.yaml (hunk 5 above)
Comment (cyclinder) [reply install/kubernetes/cilium/values.yaml]: Ok, We don't need to set the default values and let the user configure.

File: install/kubernetes/cilium/templates/cilium-configmap.yaml (hunk 2 above)
Comment (cyclinder) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml]: @joestringer I don't see anywhere where `enable-ip-masq-agent` is used,  Should we also put it in configMap `ip-masq-agnet`?

Comment (joestringer) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml]:
```
$ git grep EnableIPMasqAgent
daemon/cmd/daemon.go:   } else if option.Config.EnableIPMasqAgent {
daemon/cmd/daemon.go:   if option.Config.EnableIPMasqAgent {
daemon/cmd/daemon_main.go:      flags.Bool(option.EnableIPMasqAgent, false, "Enable BPF ip-masq-agent")
daemon/cmd/daemon_main.go:      option.BindEnv(option.EnableIPMasqAgent)
daemon/cmd/daemon_main.go:      if option.Config.EnableIPMasqAgent {
daemon/cmd/datapath.go: if option.Config.EnableIPv4 && option.Config.EnableIPMasqAgent {
daemon/cmd/status.go:           s.IPMasqAgent = option.Config.EnableIPMasqAgent
pkg/datapath/linux/config/config.go:                    if option.Config.EnableIPMasqAgent {
pkg/datapath/maps/map.go:       if !option.Config.EnableIPMasqAgent {
pkg/option/config.go:   // EnableIPMasqAgent enables BPF ip-masq-agent
pkg/option/config.go:   EnableIPMasqAgent = "enable-ip-masq-agent"
pkg/option/config.go:   EnableIPMasqAgent          bool
pkg/option/config.go:   c.EnableIPMasqAgent = viper.GetBool(EnableIPMasqAgent)
```

Comment (joestringer) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml]:
```
$ git grep EnableIPMasqAgent
daemon/cmd/daemon.go:   } else if option.Config.EnableIPMasqAgent {
daemon/cmd/daemon.go:   if option.Config.EnableIPMasqAgent {
daemon/cmd/daemon_main.go:      flags.Bool(option.EnableIPMasqAgent, false, "Enable BPF ip-masq-agent")
daemon/cmd/daemon_main.go:      option.BindEnv(option.EnableIPMasqAgent)
daemon/cmd/daemon_main.go:      if option.Config.EnableIPMasqAgent {
daemon/cmd/datapath.go: if option.Config.EnableIPv4 && option.Config.EnableIPMasqAgent {
daemon/cmd/status.go:           s.IPMasqAgent = option.Config.EnableIPMasqAgent
pkg/datapath/linux/config/config.go:                    if option.Config.EnableIPMasqAgent {
pkg/datapath/maps/map.go:       if !option.Config.EnableIPMasqAgent {
pkg/option/config.go:   // EnableIPMasqAgent enables BPF ip-masq-agent
pkg/option/config.go:   EnableIPMasqAgent = "enable-ip-masq-agent"
pkg/option/config.go:   EnableIPMasqAgent          bool
pkg/option/config.go:   c.EnableIPMasqAgent = viper.GetBool(EnableIPMasqAgent)
```

**Comment by joestringer** [comment]
I'm a bit confused what the real proposal is here. I suspect we just need some documentation about how to use the feature, not Helm changes.

I believe you should be able to do:

```
helm install ... --namespace kube-system --set ipMasqAgent.enabled=true
```

And then create a new ConfigMap in `kube-system` namespace with the complete ipmasq agent configuration, whatever this is generating:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: ip-masq-agent
  namespace: {{ .Release.Namespace }}
data:
  enable-ip-masq-agent: "true"
  config: |-
{{ toJson .Values.ipMasqAgent.config | indent 4 }}
{{- end }}
```

If we put this into the Helm charts, then my concern is that users who use this feature and upgrade will have their settings overwritten. I'm not sure if it even makes sense for Cilium's Helm configuration to control this configmap?

**Comment by joestringer** [comment]
I'm a bit confused what the real proposal is here. I suspect we just need some documentation about how to use the feature, not Helm changes.

I believe you should be able to do:

```
helm install ... --namespace kube-system --set ipMasqAgent.enabled=true
```

And then create a new ConfigMap in `kube-system` namespace with the complete ipmasq agent configuration, whatever this is generating:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: ip-masq-agent
  namespace: {{ .Release.Namespace }}
data:
  enable-ip-masq-agent: "true"
  config: |-
{{ toJson .Values.ipMasqAgent.config | indent 4 }}
{{- end }}
```

If we put this into the Helm charts, then my concern is that users who use this feature and upgrade will have their settings overwritten. I'm not sure if it even makes sense for Cilium's Helm configuration to control this configmap?

**Comment by joestringer** [comment]
One option is that we could include an example copy of this configmap in `examples/kubernetes/ip-masq-agent`, and then in the docs tell people to edit & apply that YAML directly..?

File: install/kubernetes/cilium/templates/cilium-configmap.yaml [hunk 6]
```diff
@@ -894,3 +891,17 @@ data:
 {{- end }}
 
 {{- end }}
+
+---
+{{- if and .Values.ipMasqAgent .Values.ipMasqAgent.enabled }}
```
Comment (brb) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml]: What will happen if the ip-masq-agent ConfigMap already exist? Won't it get overriden which might be unexpected by users? I think we should extend the if-condition to check whether `.Values.ipMasqAgent.config` is set.

File: install/kubernetes/cilium/templates/cilium-configmap.yaml [hunk 7]
```diff
@@ -894,3 +891,17 @@ data:
...
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: ip-masq-agent
+  namespace: {{ .Release.Namespace }}
+data:
+  enable-ip-masq-agent: "true"
```
Comment (brb) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml]: Why is it needed?

**Review by brb — CHANGES_REQUESTED** [review]

**Comment by alphabet5** [comment]
@joestringer the specific use-case I have is to eliminate a separate manifest for the configmap that isn't tied to a deployment. Having all of ciliums components managed through a single flux helmdeployment would simplify my deployment repo a bit. 

I think wrapping the whole configmap in an if statement, and leaving .Values.ipMasqAgent.config commented out would keep existing functionality the same. Then by default it's managed separately as it is now, but can be created with the helm chart.

something like

```yaml
{{- if and .Values.ipMasqAgent .Values.ipMasqAgent.enabled .Values.ipMasqAgent.config }}
apiVersion: v1
kind: ConfigMap
metadata:
  name: ip-masq-agent
  namespace: {{ .Release.Namespace }}
data:
  config: |
{{ toJson .Values.ipMasqAgent.config | indent 4 }}
{{- end }}
```

It's not a huge deal, but could make things a bit nicer.

**Comment by cyclinder** [comment]
I'm sorry  for not replying for a long time, I've been busy with work these days..

File: install/kubernetes/cilium/templates/cilium-configmap.yaml (hunk 6 above)
Comment (cyclinder) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml]: > I'm a bit confused what the core proposal is here. I suspect we just need some documentation about how to use the feature, not Helm changes.
> 
> I believe you should be able to do:
> 
> ```
> helm install ... --namespace kube-system --set ipMasqAgent.enabled=true
> ```
> 
> And then create a new ConfigMap in `kube-system` namespace with the complete ipmasq agent configuration, whatever this is generating:
> 
> ```
> apiVersion: v1
> kind: ConfigMap
> metadata:
>   name: ip-masq-agent
>   namespace: {{ .Release.Namespace }}
> data:
>   enable-ip-masq-agent: "true"
>   config: |-
> {{ toJson .Values.ipMasqAgent.config | indent 4 }}
> {{- end }}
> ```
> 
> If we put this into the Helm charts, then my concern is that users who use this feature and upgrade will have their settings overwritten. I'm not sure if it even makes sense for Cilium's Helm configuration to control this configmap?

@joestringer  Sorry to too long time no reply! Refer to @alphabet5 's comments.  I think his approach avoids the problem of  configuration override.  leaving `.Values.ipMasqAgent.config` commented out by default. 

When install  with `--set ipMasqAgent.enabled=true --set ipMasqAgent.config.masqLinkLocal=true --set "ipMasqAgent.config.nonMasqueradeCIDRs={1.1.1.1/32 ,2.2.2.2/32}" `,It will creates the configMap, which will not be overwritten even if the user updates.

File: install/kubernetes/cilium/templates/cilium-configmap.yaml (hunk 7 above)
Comment (cyclinder) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml]: Oh sorry, It's my bad..

**Review by brb — CHANGES_REQUESTED** [review]
Could you extend `Documentation/concepts/networking/masquerading.rst` to show how to use the new way of configuring the ipmasq?

File: Documentation/concepts/networking/masquerading.rst [hunk 8]
```diff
@@ -114,7 +114,23 @@ In addition, if the ``masqLinkLocal`` is not set or set to false, then
 The agent uses Fsnotify to track updates to the configuration file, so the original
 ``resyncInterval`` option is unnecessary.
 
-The example below shows how to configure the agent via :term:`ConfigMap` and to verify it:
+The example below shows how to configure the agent When you install cilium via :term:`helm install` and to verify it:
+
+.. code-block:: shell-session
```
Comment (brb) [inline Documentation/concepts/networking/masquerading.rst]: I would add this section after the installation via rfc1918.yaml, and then phrase it like:

> Alternatively, you can pass `--set ipMasqAgent.config.nonMasqueradeCIDRs='{172.16.0.0/12,192.168.0.0/16}'` and  `--set ipMasqAgent.config.masqLinkLocal=false` when installing Cilium via Helm to configure the ip-masq-agent as above.

Comment (cyclinder) [reply Documentation/concepts/networking/masquerading.rst]: Great! Thanks @brb

**Comment by joestringer** [comment]
The proposal looks reasonable to me. Should we rework the `ipMasqAgent` test in `test/k8s/datapath_configuration.go` to use this mechanism, so that we can test that the helm configuration works correctly?

**Comment by cyclinder** [comment]
> The proposal looks reasonable to me. Should we rework the `ipMasqAgent` test in `test/k8s/datapath_configuration.go` to use this mechanism, so that we can test that the helm configuration works correctly?

Hi @joestringer ,Agree this. But I would prefer to do this in another PR, Is this ok?

**Review by joestringer — APPROVED** [review]

File: Documentation/concepts/networking/masquerading.rst [hunk 9]
```diff
@@ -136,6 +136,10 @@ The example below shows how to configure the agent via :term:`ConfigMap` and to
     172.16.0.0/12
     192.168.0.0/16
 
+Alternatively, you can pass ``--set ipMasqAgent.config.nonMasqueradeCIDRs='{10.0.0.0/8,172.16.0.0/12,192.168.0.0/16}'``
+and ``--set ipMasqAgent.config.masqLinkLocal=false`` when installing Cilium via Helm to
+configure the via :term: `ip-masq-agent` as above.
```
Comment (joestringer) [inline Documentation/concepts/networking/masquerading.rst]: As per the CI docs job failure, I don't think that `ip-masq-agent` is defined as a term currently. You will need to either define the term or just quote the phrase like this:
```suggestion
configure the via ``ip-masq-agent`` as above.
```

**Comment by joestringer** [comment]
I think that the ConformanceIngress failure is likely because the branch isn't rebased against master, that one can be ignored.

**Comment by cyclinder** [comment]
Thanks @joestringer ,I am now fixing it. Does this PR need some kind of label like `/kind feature` or something?  I found that I set it, but it didn't work :)

**Comment by joestringer** [comment]
No need, I've set the `release-note/minor` to indicate that this is a user-facing change but not a major new feature. It should be good :+1:

**Review by joestringer — APPROVED** [review]

File: Documentation/concepts/networking/masquerading.rst [hunk 10]
```diff
@@ -130,6 +130,10 @@ The example below shows how to configure the agent via :term:`ConfigMap` and to
     172.16.0.0/12
     192.168.0.0/16
 
+Alternatively, you can pass ``--set ipMasqAgent.config.nonMasqueradeCIDRs='{10.0.0.0/8,172.16.0.0/12,192.168.0.0/16}'``
+and ``--set ipMasqAgent.config.masqLinkLocal=false`` when installing Cilium via Helm to
+configure the via  ``ip-masq-agent`` as above.
```
Comment (brb) [inline Documentation/concepts/networking/masquerading.rst]: Nit: `via` is not needed here.

**Comment by brb** [comment]
> Agree this. But I would prefer to do this in another PR, Is this ok?

The following diff should do the job (minus the TODO):

```
diff --git a/test/k8s/datapath_configuration.go b/test/k8s/datapath_configuration.go
index 8077da4ee1..d608fe23cb 100644
--- a/test/k8s/datapath_configuration.go
+++ b/test/k8s/datapath_configuration.go
@@ -7,8 +7,6 @@ import (
        "context"
        "fmt"
        "net"
-       "os"
-       "path/filepath"
        "regexp"
        "strconv"
        "strings"
@@ -442,18 +440,9 @@ var _ = Describe("K8sDatapathConfig", func() {
                        tmpConfigMapDirPath string
                        tmpConfigMapPath    string
                        tmpConfigYAMLPath   string
+                       nodeIP              string
                )

-               installConfig := func(cidrsInYaml string) {
-                       kubectl.ExecMiddle(fmt.Sprintf("echo 'nonMasqueradeCIDRs:\n%s' > %s",
-                               cidrsInYaml, tmpConfigMapPath)).ExpectSuccess()
-                       kubectl.CreateResource("configmap",
-                               fmt.Sprintf("ip-masq-agent --from-file=%s --namespace=%s -o yaml --dry-run=client > %s",
-                                       tmpConfigMapDirPath, helpers.CiliumNamespace, tmpConfigYAMLPath)).
-                               ExpectSuccess("Failed to create ip-masq-agent configmap file")
-                       kubectl.ApplyDefault(tmpConfigYAMLPath).ExpectSuccess("Failed to apply configmap")
-               }
-
                BeforeAll(func() {
                        // Deploy echoserver on the node which does not run Cilium to test
                        // BPF masquerading. The pod will run in the host netns, so no CNI
@@ -463,72 +452,44 @@ var _ = Describe("K8sDatapathConfig", func() {
                        res.ExpectSuccess()
                        tmpEchoPodPath = strings.Trim(res.Stdout(), "\n")
                        kubectl.ExecMiddle(fmt.Sprintf("sed 's/NODE_WITHOUT_CILIUM/%s/' %s > %s",
-                               helpers.GetFirstNodeWithoutCilium(), echoPodPath, tmpEchoPodPath)).ExpectSuccess()
+                               "k8s3", echoPodPath, tmpEchoPodPath)).ExpectSuccess()
                        kubectl.ApplyDefault(tmpEchoPodPath).ExpectSuccess("Cannot install echoserver application")
                        Expect(kubectl.WaitforPods(helpers.DefaultNamespace, "-l name=echoserver-hostnetns",
                                helpers.HelperTimeout)).Should(BeNil())

-                       // Setup ip-masq-agent configmap dir
-                       res = kubectl.ExecMiddle("mktemp -d")
-                       res.ExpectSuccess()
-                       tmpConfigMapDirPath = strings.Trim(res.Stdout(), "\n")
-                       tmpConfigMapPath = filepath.Join(tmpConfigMapDirPath, "config")
-                       res = kubectl.ExecMiddle("mktemp")
-                       res.ExpectSuccess()
-                       tmpConfigYAMLPath = strings.Trim(res.Stdout(), "\n")
-
-                       // Deploy empty ip-masq-agent config to prevent the ipmasq agent from
-                       // adding the default nonMasq CIDRs which include the echoserver's
-                       // node IP. This is needed, as the first test case expects the request
-                       // to be masqueraded.
-                       installConfig("")
-               })
-
-               AfterEach(func() {
-                       // Don't remove so that the default nonMasq CIDRs are not installed
-                       installConfig("")
+                       var err error
+                       nodeIP, err = kubectl.GetNodeIPByLabel(kubectl.GetFirstNodeWithoutCiliumLabel(), false)
+                       Expect(err).Should(BeNil())
                })

                AfterAll(func() {
-                       kubectl.DeleteResource("configmap", fmt.Sprintf("ip-masq-agent --namespace=%s", helpers.CiliumNamespace))
-
                        if tmpEchoPodPath != "" {
                                kubectl.Delete(tmpEchoPodPath)
                        }
-
-                       for _, path := range []string{tmpEchoPodPath, tmpConfigMapPath, tmpConfigMapDirPath, tmpConfigYAMLPath} {
-                               if path != "" {
-                                       os.Remove(path)
-                               }
-                       }
                })

                testIPMasqAgent := func() {
-                       // Check that requests to the echoserver from client pods are masqueraded.
-                       nodeIP, err := kubectl.GetNodeIPByLabel(kubectl.GetFirstNodeWithoutCiliumLabel(), false)
-                       Expect(err).Should(BeNil())
+                       // Check that connections from the client pods are not masqueraded
                        Expect(testPodHTTPToOutside(kubectl,
-                               fmt.Sprintf("http://%s:80", nodeIP), true, false, false)).Should(BeTrue(),
+                               fmt.Sprintf("http://%s:80", nodeIP), false, true, false)).Should(BeTrue(),
                                "Connectivity test to http://%s failed", nodeIP)

-                       // Update ip-masq-agent config to prevent masquerading to the node IP
-                       // which is running the echoserver.
-                       installConfig(fmt.Sprintf("- %s/32", nodeIP))
-
+                       // TODO remove nonMasqueradeCIDRs from the ConfigMap
                        // Wait until the ip-masq-agent config update is handled by the agent
                        time.Sleep(90 * time.Second)

-                       // Check that connections from the client pods are not masqueraded
+                       // Check that requests to the echoserver from client pods are masqueraded.
                        Expect(testPodHTTPToOutside(kubectl,
-                               fmt.Sprintf("http://%s:80", nodeIP), false, true, false)).Should(BeTrue(),
+                               fmt.Sprintf("http://%s:80", nodeIP), true, false, false)).Should(BeTrue(),
                                "Connectivity test to http://%s failed", nodeIP)
                }

                It("DirectRouting", func() {
                        deploymentManager.DeployCilium(map[string]string{
-                               "ipMasqAgent.enabled":  "true",
-                               "tunnel":               "disabled",
-                               "autoDirectNodeRoutes": "true",
+                               "ipMasqAgent.enabled":                   "true",
+                               "ipMasqAgent.config.nonMasqueradeCIDRs": fmt.Sprintf("{%s/32}", nodeIP),
+                               "tunnel":                                "disabled",
+                               "autoDirectNodeRoutes":                  "true",
                        }, DeployCiliumOptionsAndDNS)

                        testIPMasqAgent()
@@ -536,8 +497,9 @@ var _ = Describe("K8sDatapathConfig", func() {

                It("VXLAN", func() {
                        deploymentManager.DeployCilium(map[string]string{
-                               "ipMasqAgent.enabled": "true",
-                               "tunnel":              "vxlan",
+                               "ipMasqAgent.enabled":                   "true",
+                               "ipMasqAgent.config.nonMasqueradeCIDRs": fmt.Sprintf("{%s/32}", nodeIP),
+                               "tunnel":                                "vxlan",
                        }, DeployCiliumOptionsAndDNS)

                        testIPMasqAgent()
```

**Review by brb — CHANGES_REQUESTED** [review]
Minor nit.

File: Documentation/concepts/networking/masquerading.rst (hunk 10 above)
Comment (cyclinder) [reply Documentation/concepts/networking/masquerading.rst]: Thanks, Done.

**Comment by cyclinder** [comment]
> The TODO needs to be done in this PR if you decide to change the tests. Otherwise, the tests will fail.

Ok, I still prefer to make this change in another PR :), Since it may take some time, I hope this PR can be merged in first.

**Comment by joestringer** [comment]
It seems to me like 90% of the test changes are already done in the diff above, the configuration just needs to be removed. Is that more than just a `kubectl delete ...` command? It would be nice to cover this in the tests when this lands, so that we can confirm that it works and avoid regressions.

**Comment by cyclinder** [comment]
> It seems to me like 90% of the test changes are already done in the diff above, the configuration just needs to be removed. Is that more than just a `kubectl delete ...` command? It would be nice to cover this in the tests when this lands, so that we can confirm that it works and avoid regressions.

@joestringer @brb  Thanks for your suggestions. Now all the changes in the PR.

I don't think we can delete the configMap directly here, After my testing, the configMap mounted by the agent pod does not changed, so we should use `patch` here, but I'm not sure which namespace the configMap is created in, I guess it is `helpers. LogGathererNamespace`?

**Comment by brb** [comment]
> After my testing, the configMap mounted by the agent pod does not changed, so we should use patch here, but I'm not sure which namespace the configMap is created in, I guess it is helpers. LogGathererNamespace?

Yes, you can patch it. Please use `helpers.CiliumNamespace`.

**Comment by cyclinder** [comment]
> helpers.CiliumNamespace

Okay, Thanks.

File: install/kubernetes/cilium/templates/cilium-configmap.yaml [hunk 11]
```diff
@@ -946,3 +946,15 @@ data:
 {{- end }}
 
 {{- end }}
+
+---
+{{- if and .Values.ipMasqAgent .Values.ipMasqAgent.enabled .Values.ipMasqAgent.config }}
```
Comment (kaworu) [inline install/kubernetes/cilium/templates/cilium-configmap.yaml]: Go template will evaluate all operands of `and` depending on the version (used to compile Helm), so checking for `.Values.ipMasqAgent` won't help in that case (e.g. if it's set to `~`). At that point, I would suggest to rely on `.Values.ipMasqAgent.enabled` only.

```suggestion
{{- if and .Values.ipMasqAgent.enabled .Values.ipMasqAgent.config }}
```

**Review by kaworu — CHANGES_REQUESTED** [review]
Minor comment, other than that Helm related changes LGTM.

File: install/kubernetes/cilium/templates/cilium-configmap.yaml (hunk 11 above)
Comment (cyclinder) [reply install/kubernetes/cilium/templates/cilium-configmap.yaml]: Thanks for your point! Done.

**Review by sayboras — APPROVED** [review]
LGTM on helm changes, thanks 🥇

**Review by kaworu — APPROVED** [review]
Thank you!

**Comment by cyclinder** [comment]
Thanks @kaworu @sayboras ，Hi @joestringer ， Now this PR can be merge？

**Comment by joestringer** [comment]
At a glance it looked like @brb's latest feedback was addressed, so I just triggered CI to double-check that it's passing. Assuming that's all good and @brb didn't have further feedback, this should be good to merge.

File: test/k8s/datapath_configuration.go [hunk 12]
```diff
@@ -463,81 +449,53 @@ var _ = Describe("K8sDatapathConfig", func() {
 			res.ExpectSuccess()
 			tmpEchoPodPath = strings.Trim(res.Stdout(), "\n")
 			kubectl.ExecMiddle(fmt.Sprintf("sed 's/NODE_WITHOUT_CILIUM/%s/' %s > %s",
-				helpers.GetFirstNodeWithoutCilium(), echoPodPath, tmpEchoPodPath)).ExpectSuccess()
+				"k8s3", echoPodPath, tmpEchoPodPath)).ExpectSuccess()
```
Comment (brb) [inline test/k8s/datapath_configuration.go]: Why this change?

Comment (cyclinder) [reply test/k8s/datapath_configuration.go]: I'm not sure. This change comes from your suggestion. https://github.com/cilium/cilium/pull/20137#issuecomment-1187125760, Is the CI failure related to this change?

**Comment by brb** [comment]
@cyclinder Thanks. Could you rebase your code against the latest Cilium's master. It should fix some GH action failures. Once it's done, I will rerun the CI.

**Comment by cyclinder** [comment]
> @cyclinder Thanks. Could you rebase your code against the latest Cilium's master. It should fix some GH action failures. Once it's done, I will rerun the CI.

Thanks @brb , I've already rebased with the latest master branch.

**Comment by cyclinder** [comment]
It seems to the [CI failures](https://github.com/cilium/cilium/actions/runs/2860336314)  has no relation with these changes ?

**Review by brb — APPROVED** [review]
Thanks!
